### PR TITLE
[nearley] adapt typings to v2.11

### LINKED ISSUE #19255: nearley types are wrong
I'm the maintainer of nearley, one of the libraries for which this library provides types. I've gotten a [bug report on nearley](https://github.com/Hardmath123/nearley/issues/276) that we traced to an error in this library. Specifically, the following line is incorrect:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/516ed213ca51babea056b771e431109709f1829a/types/nearley/index.d.ts#L62

Other lines might also be wrong, since the nearley team didn't write that file.

(cc @deltaidea @tjvr @notsonotso. @deltaidea is the one who knows how to fix this.)

### PR DESCRIPTION
**Pull request by BendingBender**
Fixes #19255.

@Hardmath123, @notsonotso, @deltaidea: Does this look like it fixes your errors?

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://nearley.js.org
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by notsonotso** [comment]
@BendingBender I haven't tested this thoroughly, but the obvious errors appears to be gone. 

Would be great if someone could write a small test suit to make sure the types work as new Nearley versions are released.

Great work!

**Comment by Hardmath123** [comment]
CC @tjvr as well

File: types/nearley/index.d.ts [hunk 1]
```diff
@@ -35,10 +25,22 @@ export class Parser {
      */
     results: any[];
 
+    constructor(rules: CompiledRules, start: string, options?: ParserOptions);
```
Comment (tjvr) [inline types/nearley/index.d.ts]: This constructor is deprecated, can we remove it?

File: types/nearley/index.d.ts [hunk 2]
```diff
@@ -47,26 +49,46 @@ export interface ParserOptions {
...
+    postprocess?(data: any[], reference: number, wantedBy: any): void;
+
+    constructor(name: string, symbols: any[], postprocess?: (data: any[], reference: number, wantedBy: any) => void);
+
+    toString(withCursorAt?: number): string;
 }
 
 export class Grammar {
-    constructor(rules: Rule[], start: string);
+    static fromCompiled(rules: CompiledRules, start?: string): Grammar;
```
Comment (tjvr) [inline types/nearley/index.d.ts]: Prefer `fromCompiled(rules: CompiledRules)`: the `start` option is deprecated, and retained only for backwards compat.

File: types/nearley/index.d.ts [hunk 3]
```diff
@@ -47,26 +49,46 @@ export interface ParserOptions {
...
 
 export class Rule {
-    constructor(name: any, symbols: any, postprocess: any);
-    toString(withCursorAt: any): any;
     static highestId: number;
+
+    id: number;
+    name: string;
+    symbols: any[];
+    postprocess?(data: any[], reference: number, wantedBy: any): void;
```
Comment (tjvr) [inline types/nearley/index.d.ts]: I'm not sure why `wantedBy` is here: the argument is always the value `Parser.fail`, which is typed as `static fail: {};` above. Looks like a typo to me. :)

**Comment by tjvr** [comment]
> Would be great if someone could write a small test suit to make sure the types work as new Nearley versions are released.

You're welcome to volunteer! :-)

**Comment by notsonotso** [comment]
@tjvr no time, unfortunately

**Comment by Hardmath123** [comment]
> You're welcome to volunteer! :-)

And @tjvr, that's the Brianest thing you've said all week. :P

File: types/nearley/index.d.ts (hunk 1 above)
Comment (BendingBender) [reply types/nearley/index.d.ts]: Didn't know about that, it's not mentioned in the source. Will remove.

File: types/nearley/index.d.ts (hunk 2 above)
Comment (BendingBender) [reply types/nearley/index.d.ts]: Didn't know about that, it's not mentioned in the source. Will remove.

**Comment by notsonotso** [comment]
Great work guys, makes it a lot easier to use Nearley from Typescript!

**Comment by Hardmath123** [comment]
Once this is merged, let me know and I'll add it to the "Other tools" section of the documentation so that others can find it. https://nearley.js.org/docs/tooling#other-tools

File: types/nearley/index.d.ts [hunk 4]
```diff
@@ -35,10 +25,21 @@ export class Parser {
      */
     results: any[];
```
Comment (BendingBender) [inline types/nearley/index.d.ts]: @tjvr Any idea what is the type of this property?

**Comment by andy-ms** [comment]
Thanks!

**Comment by andy-ms** [comment]
Un-deprecated `@types/nearley`.

**Pull request merged**